src/main/java/com/powsybl/openloadflow/ac/outerloop/AcEmulationOuterLoop.java

```diff
@@ -0,0 +1,149 @@
+            if (acEmulationControl.getFeedingSide() == TwoSides.ONE) {
+                if (hvdc.getP1().eval() > acEmulationControl.getPMaxFromCS1toCS2()) {
+                    // Switch mode
+                    LOGGER.trace("Bound Hvdc flow to Pmax from CS1 to CS2 for Hvdc: {}", hvdcId);
+                    contextData.incrementModeSwitchCount(hvdcId);
+                    hvdc.updateAcEmulationStatus(LfHvdc.AcEmulationControl.AcEmulationStatus.BOUNDED);
+                    if (contextData.getModeSwitchCount(hvdcId) == MAX_MODE_SWITCH) {
+                        LOGGER.debug("Two many mode switches (flow blocked to Pmax from CS1 to CS2) for Hvdc: {}", hvdcId);
+                    }
+                    return true;
+                }
+            } else {
+                if (hvdc.getP2().eval() > acEmulationControl.getPMaxFromCS2toCS1()) {
+                    // Switch mode
+                    LOGGER.trace("Bound Hvdc flow to Pmax from CS2 to CS1 for Hvdc: {}", hvdcId);
+                    contextData.incrementModeSwitchCount(hvdcId);
+                    hvdc.updateAcEmulationStatus(LfHvdc.AcEmulationControl.AcEmulationStatus.BOUNDED);
+                    if (contextData.getModeSwitchCount(hvdcId) == MAX_MODE_SWITCH) {
+                        LOGGER.debug("Two many mode switches (flow blocked to Pmax from CS2 to CS1) for Hvdc: {}", hvdcId);
+                    }
+                    return true;
+                }
+            }
+        }
+
+        // Check for mode switch between BOUNDED and FREE
+        if (acEmulationControl.getAcEmulationStatus() == LfHvdc.AcEmulationControl.AcEmulationStatus.BOUNDED) {
+            if (acEmulationControl.getFeedingSide() == TwoSides.ONE) {
+                if (computeRawP1(hvdc) < acEmulationControl.getPMaxFromCS1toCS2()) {
+                    // Switch mode
+                    LOGGER.trace("Set free the Ac Emulation mode for Hvdc: {}", hvdcId);
+                    hvdc.updateAcEmulationStatus(LfHvdc.AcEmulationControl.AcEmulationStatus.FREE);
+                    return true;
+                }
+            } else {
+                if (computeRawP2(hvdc) < acEmulationControl.getPMaxFromCS2toCS1()) {
+                    // Switch mode
+                    LOGGER.trace("Set free the Ac Emulation mode for Hvdc: {}", hvdcId);
+                    hvdc.updateAcEmulationStatus(LfHvdc.AcEmulationControl.AcEmulationStatus.FREE);
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public OuterLoopResult check(AcOuterLoopContext context, ReportNode reportNode) {
+        OuterLoopStatus status = OuterLoopStatus.STABLE;
+        ContextData contextData = (ContextData) context.getData();
+
+        for (LfHvdc hvdc : context.getNetwork().getHvdcs()) {
+            if (!hvdc.isAcEmulation() || hvdc.getBus1().isDisabled() || hvdc.getBus2().isDisabled() || hvdc.isDisabled()) {
+                continue;
```

Comment:
What you really want to know here is whether the hvdc link is active. That is if the hvdc in AC emulation mode and the  P1 and P2 equations terms are active.
This condition here looks like a code duplication and may create problems in the future if we need to change the conditions and notifications that activate/deactivate P1 and P2 for LfHDVC.
Why not making getP1 and getP2 returning an EquationTerm instead of an evaluable and checking their active status ?
That would ensure consistency.

Reply:
Ok I will try that